File: editors/jsfcstm/src/diagnostics/analyzers/use-def.ts
import {
    BinaryOp,
    ConditionalOp,
    Expr,
    IfBlock,
    Operation,
    OperationStatement,
    StateMachine,
    UFunc,
    UnaryOp,
    Variable,
} from '../../model/runtime';

export class UseDefGraph {
    readonly edges: Array<[string, string]>;
    readonly dependenciesByTarget: Record<string, string[]>;

    constructor(edges: Array<[string, string]>) {
        this.edges = [...edges].sort(comparePair);
        const deps: Record<string, string[]> = {};
        for (const [source, target] of this.edges) {
            deps[target] = [...(deps[target] ?? []), source];
        }
        this.dependenciesByTarget = {};
        for (const target of Object.keys(deps).sort()) {
            this.dependenciesByTarget[target] = deps[target];
        }
    }

    dependenciesOf(target: string): string[] {
        return this.dependenciesByTarget[target] ?? [];
    }

    affectingVariables(directVariables: Iterable<string>): string[] {
        const direct = new Set(directVariables);
        const seen = new Set(direct);
        const out = new Set<string>();
        const queue = Array.from(direct);
        while (queue.length > 0) {
            const target = queue.shift()!;
            for (const source of this.dependenciesOf(target)) {
                if (seen.has(source)) continue;
                seen.add(source);
                out.add(source);
                queue.push(source);
            }
        }
        return Array.from(out).sort();
    }
}

export function buildUseDefGraph(machine: StateMachine): UseDefGraph {
    const edges = new Set<string>();

    for (const state of machine.allStates) {
        for (const collection of [state.onEnters, state.onDurings, state.onExits, state.onDuringAspects]) {
            for (const action of collection) {
                if (action.isAbstract) continue;
                walkBlock(action.operations ?? [], new Set(), edges);
            }
        }
        for (const transition of state.transitions) {
            walkBlock(transition.effects ?? [], new Set(), edges);
        }
    }

    return new UseDefGraph(Array.from(edges).map(splitEdge));
}

export function collectExprVariables(expr: Expr): string[] {
    const out: string[] = [];
    walkExprCollect(expr, out);
    return Array.from(new Set(out));
}

function walkBlock(
    statements: OperationStatement[],
    enclosingCondVars: Set<string>,
    edges: Set<string>,
): void {
    for (const stmt of statements) {
        if (stmt instanceof Operation) {
            const deps = new Set([...collectExprVariables(stmt.expr), ...enclosingCondVars]);
            for (const source of deps) {
                edges.add(joinEdge(source, stmt.varName));
            }
            continue;
        }
        if (stmt instanceof IfBlock) {
            const accumulated = new Set<string>();
            for (const branch of stmt.branches) {
                const thisCond = new Set(branch.condition ? collectExprVariables(branch.condition) : []);
                const branchEnclosing = new Set([
                    ...enclosingCondVars,
                    ...accumulated,
                    ...thisCond,
                ]);
                walkBlock(branch.statements, branchEnclosing, edges);
                for (const item of thisCond) accumulated.add(item);
            }
        }
    }
}

function walkExprCollect(expr: Expr, out: string[]): void {
    if (expr instanceof Variable) {
        out.push(expr.name);
        return;
    }
    if (expr instanceof UnaryOp || expr instanceof UFunc) {
        walkExprCollect(expr.x, out);
        return;
    }
    if (expr instanceof BinaryOp) {
        walkExprCollect(expr.x, out);
        walkExprCollect(expr.y, out);
        return;
    }
    if (expr instanceof ConditionalOp) {
        walkExprCollect(expr.cond, out);
        walkExprCollect(expr.ifTrue, out);
        walkExprCollect(expr.ifFalse, out);
    }
}

function joinEdge(source: string, target: string): string {
    return `${source}\u0001${target}`;
}

function splitEdge(edge: string): [string, string] {
    const [source, target] = edge.split('\u0001');
    return [source, target];
}

function comparePair(left: [string, string], right: [string, string]): number {
    if (left[0] < right[0]) return -1;
    if (left[0] > right[0]) return 1;
    if (left[1] < right[1]) return -1;
    if (left[1] > right[1]) return 1;
    return 0;
}

File: pyfcstm/diagnostics/analyzers/use_def.py
"""Use-def graph construction for guard-affect data-flow analysis."""

from dataclasses import dataclass
from typing import TYPE_CHECKING, Dict, Iterable, List, Set, Tuple

if TYPE_CHECKING:  # pragma: no cover - import-time type hints only
    from ...model.expr import Expr
    from ...model.model import OperationStatement, StateMachine


@dataclass(frozen=True)
class UseDefGraph:
    """Directed variable dependency graph.

    Each edge is ``(source, target)`` and means that ``target``'s value may
    depend on ``source`` through an assignment expression or an enclosing
    operation-block condition.
    """

    edges: Tuple[Tuple[str, str], ...]
    dependencies_by_target: Dict[str, Tuple[str, ...]]

    def dependencies_of(self, target: str) -> Tuple[str, ...]:
        """Return variables that may affect ``target``."""
        return self.dependencies_by_target.get(target, tuple())

    def affecting_variables(self, direct_variables: Iterable[str]) -> Tuple[str, ...]:
        """Return all variables that can flow into ``direct_variables``."""
        direct = set(direct_variables)
        seen: Set[str] = set(direct)
        out: Set[str] = set()
        queue = list(direct)
        while queue:
            target = queue.pop(0)
            for source in self.dependencies_of(target):
                if source in seen:
                    continue
                seen.add(source)
                out.add(source)
                queue.append(source)
        return tuple(sorted(out))


def build_use_def_graph(machine: 'StateMachine') -> UseDefGraph:
    """Build a conservative use-def graph from concrete actions/effects."""
    edges: Set[Tuple[str, str]] = set()

    for state in machine.walk_states():
        for collection in (
                state.on_enters,
                state.on_durings,
                state.on_exits,
                state.on_during_aspects,
        ):
            for action in collection:
                if getattr(action, 'is_abstract', False):
                    continue
                _walk_block(action.operations or [], set(), edges)

        for transition in state.transitions:
            _walk_block(transition.effects or [], set(), edges)

    ordered_edges = tuple(sorted(edges))
    deps: Dict[str, List[str]] = {}
    for source, target in ordered_edges:
        deps.setdefault(target, []).append(source)
    return UseDefGraph(
        edges=ordered_edges,
        dependencies_by_target={
            target: tuple(sources)
            for target, sources in sorted(deps.items())
        },
    )


def collect_expr_variables(expr: 'Expr') -> Tuple[str, ...]:
    """Return variable names read by ``expr`` in stable first-seen order."""
    out: List[str] = []
    _walk_expr_collect(expr, out)
    seen = set()
    deduped: List[str] = []
    for name in out:
        if name in seen:
            continue
        seen.add(name)
        deduped.append(name)
    return tuple(deduped)


def _walk_block(
        statements: Iterable['OperationStatement'],
        enclosing_cond_vars: Set[str],
        edges: Set[Tuple[str, str]],
) -> None:
    from ...model.model import IfBlock, Operation

    for stmt in statements:
        if isinstance(stmt, Operation):
            deps = set(collect_expr_variables(stmt.expr)) | set(enclosing_cond_vars)
            for source in deps:
                edges.add((source, stmt.var_name))
            continue
        if isinstance(stmt, IfBlock):
            accumulated: Set[str] = set()
            for branch in stmt.branches:
                this_cond = (
                    set(collect_expr_variables(branch.condition))
                    if branch.condition is not None
                    else set()
                )
                branch_enclosing = (
                    set(enclosing_cond_vars) | accumulated | this_cond
                )
                _walk_block(branch.statements, branch_enclosing, edges)
                accumulated.update(this_cond)


def _walk_expr_collect(expr: 'Expr', out: List[str]) -> None:
    from ...model.expr import (
        BinaryOp,
        ConditionalOp,
        UFunc,
        UnaryOp,
        Variable,
    )

    if isinstance(expr, Variable):
        out.append(expr.name)
        return
    if isinstance(expr, UnaryOp):
        _walk_expr_collect(expr.x, out)
        return
    if isinstance(expr, BinaryOp):
        _walk_expr_collect(expr.x, out)
        _walk_expr_collect(expr.y, out)
        return
    if isinstance(expr, ConditionalOp):
        _walk_expr_collect(expr.cond, out)
        _walk_expr_collect(expr.if_true, out)
        _walk_expr_collect(expr.if_false, out)
        return
    if isinstance(expr, UFunc):
        _walk_expr_collect(expr.x, out)
